src/State/Provider/ReadProvider.php

```diff
@@ -87,6 +90,47 @@ public function provide(Operation $operation, array $uriVariables = [], array $c
             throw new NotFoundHttpException('Not Found');
         }
 
+        if ($operation->getUriVariables()) {
+            foreach ($operation->getUriVariables() as $key => $uriVariable) {
+                if (!$uriVariable instanceof Link || !$uriVariable->getSecurity()) {
+                    continue;
+                }
+
+                $relationClass = $uriVariable->getFromClass() ?? $uriVariable->getToClass();
+
+                if (!$relationClass) {
+                    continue;
+                }
+
+                $parentOperation = $this->resourceMetadataCollectionFactory
+                    ->create($relationClass)
+                    ->getOperation($operation->getExtraProperties()['parent_uri_template'] ?? null);
+                try {
+                    $relation = $this->provider->provide($parentOperation, [$uriVariable->getIdentifiers()[0] => $request?->attributes->all()[$key]], $context);
+                } catch (ProviderNotFoundException) {
+                    $relation = null;
+                }
+                if (!$relation) {
+                    throw new NotFoundHttpException('Not Found');
+                }
+
+                try {
+                    $securityObjectName = $uriVariable->getSecurityObjectName();
+
+                    if (!$securityObjectName) {
+                        $securityObjectName = $uriVariable->getToProperty() ?? $uriVariable->getFromProperty();
+                    }
+
+                    if (!$securityObjectName) {
+                        continue;
+                    }
+                    $request?->attributes->set($securityObjectName, $relation);
+                } catch (InvalidIdentifierException|InvalidUriVariableException $e) {
+                    throw new NotFoundHttpException('Invalid identifier value or configuration.', $e);
+                }
+            }
+        }
+
```

Comment:
I think that this may have an impact over performances, it's quite magical and you must be sure to set `parent_uri_template` or to have the correct first operation on the linked resource. Also, by the looks of it I'm wondering if we couldn't just check out the `$data->($uriVariables->getFromProperty()}` instead of calling a read provider on that as it may fetch more then we want no? 

Anyways, the way it could be implemented if calling the provider is needed:

- create a `LinkReadProvider` (or another name not sure ^^) that decorates the `api_platform.state_provider.read` (take example on `api_platform.state_provider.access_checker`)
- this way you can make this `opt-in` add a configuration option that enables link security

Reply:
I just remembered why that was there:

If $data is a collection you have no way of actually getting the parent element except trying to fetch it off the DB or other things. So I'd argue using a provider to grab the related field is required.

Sadly that also doesn't work anymore when decorating the read provider, basically making me unable to get the owning object.

Do you by chance know another way I could get that object?

Reply:
yes the `parent_uri_template` is really useful for that. 

> Sadly that also doesn't work anymore when decorating the read provider, basically making me unable to get the owning object. 

You got it by calling `$this->decorated->provide()` no? or else you can always use the `$request->attributes` to keep data during a request.